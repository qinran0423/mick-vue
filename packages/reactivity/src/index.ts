export { ref, proxyRefs, isRef } from "./ref"
export { ReactiveEffect, effect } from "./effect"
export { reactive, isReactive, shallowReadonly } from "./reactive"
