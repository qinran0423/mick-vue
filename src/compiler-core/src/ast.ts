
export const enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT
}