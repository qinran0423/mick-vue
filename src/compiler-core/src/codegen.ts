

export function generate(ast) {
  const context = createCodegenContext()
  const { push } = context
  push("return ")

  const functionName = "render"
  const args = ["_ctx", "_cache"]
  const signature = args.join(", ")
  push(`function ${functionName}(${signature}){`)
  push(`return`)
  genNode(ast.codegenNode, context)
  push("}")

  return {
    code: context.code
  }
}

function genNode(node: any, context) {
  const { push } = context
  push(`'${node.content}'`)
}
function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    }
  }
  return context
}

