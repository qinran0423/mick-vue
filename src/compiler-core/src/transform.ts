
export function transform(root, options = {}) {
  const context = createTransformsContext(root, options)

  traverseNode(root, context)

  createRootCodegen(root)
}

function traverseNode(node: any, context) {
  const nodeTransforms = context.nodeTransforms
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i]
    transform(node)
  }

  traverseChildren(node, context)
}

function traverseChildren(node: any, context: any) {
  const children = node.children

  if (children) {
    for (let i = 0; i < children.length; i++) {
      const node = children[i]
      traverseNode(node, context)
    }
  }
}

function createTransformsContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || []
  }

  return context

}

function createRootCodegen(root: any) {
  root.codegenNode = root.children[0]
}

