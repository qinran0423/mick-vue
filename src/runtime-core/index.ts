export { h } from './h'
export { renderSlots } from './helpers/renderSlots'
export { createTextVNode } from './vnode'
export { getCurrentInstance } from './component'
export { provide, inject } from './apiinject'
export { createRenderer } from './renderer'
export { nextTick } from './scheduler'