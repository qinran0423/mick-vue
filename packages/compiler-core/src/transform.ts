import { NodeTypes } from "./ast"
import { TO_DISPLAT_STRING } from "./transforms/runtimeHelpers"

export function transform(root, options = {}) {
  const context = createTransformsContext(root, options)

  traverseNode(root, context)

  createRootCodegen(root)

  root.helpers = [...context.helpers.keys()]
}

function traverseNode(node: any, context) {
  // 节点转换函数
  const nodeTransforms = context.nodeTransforms
  const exitFns: any = []
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i]
    // 有些转换函数会设计一个退出函数，在处理完子节点后执行
    const onExit = transform(node, context)
    if (onExit) exitFns.push(onExit)
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      // 需要导入toString辅助函数
      context.helper(TO_DISPLAT_STRING)
      break
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      // 遍历子节点
      traverseChildren(node, context)
      break
    default:
      break
  }

  // 执行转换函数返回的退出函数
  let i = exitFns.length
  while (i--) {
    exitFns[i]()
  }
}

function traverseChildren(node: any, context: any) {
  const children = node.children

  for (let i = 0; i < children.length; i++) {
    const node = children[i]
    traverseNode(node, context)
  }
}

function createTransformsContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1)
    }
  }

  return context
}

function createRootCodegen(root: any) {
  const child = root.children[0]
  if (child.type === NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode
  } else {
    root.codegenNode = root.children[0]
  }
}
