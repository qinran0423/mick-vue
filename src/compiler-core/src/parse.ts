import { NodeTypes } from "./ast"

const enum TagType {
  Start,
  End
}

export function baseParse(content: string) {

  const context = createParserContext(content)

  return createRoot(parseChildren(context, []))
}


function parseChildren(context, ancestors) {
  const nodes: any = []


  while (!isEnd(context, ancestors)) {
    let node
    const s = context.source
    if (s.startsWith("{{")) {
      node = parseInterpolation(context)
    } else if (s[0] === '<') {
      if (/[a-z]/i.test(s[1])) {
        console.log('parse element');
        node = parseElement(context, ancestors)
      }
    }

    if (!node) {
      node = parseText(context)
    }
    nodes.push(node)
  }

  console.log('nodes', nodes);


  return nodes
}

function isEnd(context, ancestors) {
  const s = context.source

  if (s.startsWith('</')) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag;
      if (startsWithEndTagOpen(s, tag)) {
        return true
      }
    }
  }

  // if (parentTag && s.startsWith(`</${parentTag}>`)) {
  //   return true
  // } else {

  // }
  return !s
}


function parseText(context: any): any {

  let endIndex = context.source.length
  let endToken = ['<', '{{']

  for (let i = 0; i < endToken.length; i++) {
    const index = context.source.indexOf(endToken[i])
    if (index !== -1 && endIndex > index) {
      endIndex = index
    }
  }



  const content = parseTextData(context, endIndex)
  console.log('content----', content);

  return {
    type: NodeTypes.TEXT,
    content
  }

}

function parseTextData(context: any, length) {
  const content = context.source.slice(0, length)

  advanceBy(context, content.length)
  return content
}

function parseElement(context, ancestors) {
  // 1.解析tag
  const element: any = parseTag(context, TagType.Start)
  ancestors.push(element)
  element.children = parseChildren(context, ancestors)
  ancestors.pop()


  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagType.End)
  } else {
    throw new Error(`缺少结束标签: ${element.tag}`)
  }


  return element
}

function startsWithEndTagOpen(source, tag) {
  return source.startsWith('</') && source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
}



function parseTag(context: any, type: TagType) {
  const match: any = /^<\/?([a-z]*)/i.exec(context.source)
  const tag = match[1]
  // 2.删除处理完成的代码
  advanceBy(context, match[0].length)
  advanceBy(context, 1)

  if (type === TagType.End) return;

  return {
    type: NodeTypes.ELEMENT,
    tag
  }
}

function parseInterpolation(context) {

  const openDelimiter = '{{'
  const closeDelimiter = '}}'

  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)

  advanceBy(context, openDelimiter.length)

  const rawContentLength = closeIndex - openDelimiter.length

  const rawContent = parseTextData(context, rawContentLength)

  const content = rawContent.trim()
  advanceBy(context, closeDelimiter.length)

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content
    }
  }
}

function advanceBy(context: any, length: number) {
  context.source = context.source.slice(length)
}

function createRoot(children) {
  return {
    children
  }
}

function createParserContext(content: string) {
  return {
    source: content
  }
}
