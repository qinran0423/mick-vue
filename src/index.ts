
// mick-vue的出口
export * from './runtime-dom'
import { baseCompile } from './compiler-core/src'
import * as runtimeDom from './runtime-dom'
import { registerRuntimeCompiler } from './runtime-dom'

function compileToFunction(template) {
  const { code } = baseCompile(template)

  const render = new Function("Vue", code)(runtimeDom)

  return render
}

registerRuntimeCompiler(compileToFunction)