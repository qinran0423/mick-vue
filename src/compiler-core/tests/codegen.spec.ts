import { generate } from "../src/codegen";
import { baseParse } from "../src/parse"
import { transform } from "../src/transform";

describe('codegen', () => {
  it('string', () => {
    const ast = baseParse('hi')
    console.log(ast);
    transform(ast)
    const { code } = generate(ast)

    // 快照
    expect(code).toMatchSnapshot()
  })
})