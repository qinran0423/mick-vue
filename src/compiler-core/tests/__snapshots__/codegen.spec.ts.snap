// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen string 1`] = `"return function render(_ctx, _cache){return'hi'}"`;
